### ISSUE TITLE
Darwin: cross compilation of glibc fails

### ISSUE THREAD

**Issue by mroi** [NixOS/nixpkgs#80754]
**Describe the bug**
I was trying to build a Linux cross compiler on Darwin, but it fails when configuring glibc.

**To Reproduce**
`nix-build -A pkgsCross.gnu64.glibc '<nixpkgs>'`

**Expected behavior**
Cross-compiled glibc should be built.

**Additional context**
I believe this is related to #77632. Commendable work has been done there, but the passing of macOS-specific linker options in `NIX_LDFLAGS` appears to be problematic for cross builds.

I am not knowledgeable enough about Nix’ cross compilation infrastructure to fully diagnose the problem, let alone suggest a fix. But I think `NIX_LDFLAGS` should not be passed to linkers targeting non-Darwin platforms, because they would not understand the `-macosx_version_min` options.

**Metadata**
 - system: `"x86_64-darwin"`
 - host os: `Darwin 19.4.0, macOS 10.15.4`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.3`
 - channels(Michael): `"nixpkgs-20.09pre214168.686274ea62c"`
 - nixpkgs: `/Users/Michael/.nix-defexpr/channels/nixpkgs`

**Comment by matthewbauer** [comment]
Yeah, I'm not sure the best way to deal with this. Perhaps we can set `NIX_BUILD_LDFLAGS` instead of `NIX_LDFLAGS`.

/cc @Ericson2314

**Comment by LnL7** [comment]
I guess that might make sense if dyld isn't used as the linker in that case. Could run and post the output of `otool -l` on a cross compiled binary?

**Comment by mroi** [comment]
I think we have to separate between binaries being hosted on the local platform, but targetting the cross platform and those already intended to run on the cross platform.

The Linux linker `i686-linux-ld` falls into the first category. It creates Linux executables, but runs on macOS. It is linked like this:

```
Michael@edr06939:~ > otool -L /nix/store/jrdxi1lcngimljm7xy101d6i9hgr7373-i686-linux-binutils-2.31.1/bin/i686-linux-ld
/nix/store/jrdxi1lcngimljm7xy101d6i9hgr7373-i686-linux-binutils-2.31.1/bin/i686-linux-ld:
	/nix/store/jrdxi1lcngimljm7xy101d6i9hgr7373-i686-linux-binutils-2.31.1/x86_64-apple-darwin/i686-linux/lib/libbfd-2.31.1.dylib (compatibility version 0.0.0, current version 0.0.0)
	/nix/store/fxvxl64g1b336ayhzsrqdcv541zpb6lx-Libsystem-osx-10.12.6/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
	/nix/store/ww5azqlhhp064brjqi7k2c6s30qsi0bg-zlib-1.2.11/lib/libz.dylib (compatibility version 1.0.0, current version 1.2.11)
	/nix/store/rb3m0xyx1l1jg2bl837w11jhm6ax1yrs-gettext-0.20.1/lib/libintl.8.dylib (compatibility version 10.0.0, current version 10.6.0)
```

`glibc` would fall in the second category. To build `glibc`, the Linux linker is invoked with macOS-specific linker flags from `NIX_LDFLAGS`, which it does not understand. Therefore, `glibc` fails in the configure step.

**Comment by LnL7** [comment]
@mroi I'm looking fir the output of the lowecase `-l` flag.

**Comment by mroi** [comment]
Sorry. Stupid me.

I’m on a different machine now, and here I have a version of the x86_64 Linux Linker. I hope that’s OK for you.

`otool -l /nix/store/ayc7d1h0alm1y2pysr2hbv183g7d6ji7-x86_64-unknown-linux-gnu-binutils-2.31.1/bin/x86_64-unknown-linux-gnu-ld`
<details>
<summary>Command Output</summary>

    /nix/store/ayc7d1h0alm1y2pysr2hbv183g7d6ji7-x86_64-unknown-linux-gnu-binutils-2.31.1/bin/x86_64-unknown-linux-gnu-ld:
    Mach header
          magic cputype cpusubtype  caps    filetype ncmds sizeofcmds      flags
     0xfeedfacf 16777223          3  0x80           2    18       2192 0x00200085
    Load command 0
          cmd LC_SEGMENT_64
      cmdsize 72
      segname __PAGEZERO
       vmaddr 0x0000000000000000
       vmsize 0x0000000100000000
      fileoff 0
     filesize 0
      maxprot 0x00000000
     initprot 0x00000000
       nsects 0
        flags 0x0
    Load command 1
          cmd LC_SEGMENT_64
      cmdsize 552
      segname __TEXT
       vmaddr 0x0000000100000000
       vmsize 0x0000000000164000
      fileoff 0
     filesize 1458176
      maxprot 0x00000007
     initprot 0x00000005
       nsects 6
        flags 0x0
    Section
      sectname __text
       segname __TEXT
          addr 0x0000000100000d70
          size 0x0000000000040113
        offset 3440
         align 2^4 (16)
        reloff 0
        nreloc 0
         flags 0x80000400
     reserved1 0
     reserved2 0
    Section
      sectname __stubs
       segname __TEXT
          addr 0x0000000100040e84
          size 0x0000000000000492
        offset 265860
         align 2^1 (2)
        reloff 0
        nreloc 0
         flags 0x80000408
     reserved1 0 (index into indirect symbol table)
     reserved2 6 (size of stubs)
    Section
      sectname __stub_helper
       segname __TEXT
          addr 0x0000000100041318
          size 0x00000000000007ae
        offset 267032
         align 2^2 (4)
        reloff 0
        nreloc 0
         flags 0x80000400
     reserved1 0
     reserved2 0
    Section
      sectname __cstring
       segname __TEXT
          addr 0x0000000100041ac6
          size 0x0000000000118403
        offset 268998
         align 2^0 (1)
        reloff 0
        nreloc 0
         flags 0x00000002
     reserved1 0
     reserved2 0
    Section
      sectname __const
       segname __TEXT
          addr 0x0000000100159ed0
          size 0x0000000000009c30
        offset 1416912
         align 2^4 (16)
        reloff 0
        nreloc 0
         flags 0x00000000
     reserved1 0
     reserved2 0
    Section
      sectname __unwind_info
       segname __TEXT
          addr 0x0000000100163b00
          size 0x00000000000004f4
        offset 1456896
         align 2^2 (4)
        reloff 0
        nreloc 0
         flags 0x00000000
     reserved1 0
     reserved2 0
    Load command 2
          cmd LC_SEGMENT_64
      cmdsize 632
      segname __DATA
       vmaddr 0x0000000100164000
       vmsize 0x0000000000007000
      fileoff 1458176
     filesize 24576
      maxprot 0x00000007
     initprot 0x00000003
       nsects 7
        flags 0x0
    Section
      sectname __nl_symbol_ptr
       segname __DATA
          addr 0x0000000100164000
          size 0x0000000000000010
        offset 1458176
         align 2^3 (8)
        reloff 0
        nreloc 0
         flags 0x00000006
     reserved1 195 (index into indirect symbol table)
     reserved2 0
    Section
      sectname __got
       segname __DATA
          addr 0x0000000100164010
          size 0x0000000000000078
        offset 1458192
         align 2^3 (8)
        reloff 0
        nreloc 0
         flags 0x00000006
     reserved1 197 (index into indirect symbol table)
     reserved2 0
    Section
      sectname __la_symbol_ptr
       segname __DATA
          addr 0x0000000100164088
          size 0x0000000000000618
        offset 1458312
         align 2^3 (8)
        reloff 0
        nreloc 0
         flags 0x00000007
     reserved1 212 (index into indirect symbol table)
     reserved2 0
    Section
      sectname __const
       segname __DATA
          addr 0x00000001001646a0
          size 0x0000000000004230
        offset 1459872
         align 2^4 (16)
        reloff 0
        nreloc 0
         flags 0x00000000
     reserved1 0
     reserved2 0
    Section
      sectname __data
       segname __DATA
          addr 0x00000001001688d0
          size 0x00000000000010b0
        offset 1476816
         align 2^4 (16)
        reloff 0
        nreloc 0
         flags 0x00000000
     reserved1 0
     reserved2 0
    Section
      sectname __bss
       segname __DATA
          addr 0x0000000100169980
          size 0x0000000000000d00
        offset 0
         align 2^4 (16)
        reloff 0
        nreloc 0
         flags 0x00000001
     reserved1 0
     reserved2 0
    Section
      sectname __common
       segname __DATA
          addr 0x000000010016a680
          size 0x0000000000000454
        offset 0
         align 2^3 (8)
        reloff 0
        nreloc 0
         flags 0x00000001
     reserved1 0
     reserved2 0
    Load command 3
          cmd LC_SEGMENT_64
      cmdsize 72
      segname __LINKEDIT
       vmaddr 0x000000010016b000
       vmsize 0x000000000000c814
      fileoff 1482752
     filesize 51220
      maxprot 0x00000007
     initprot 0x00000001
       nsects 0
        flags 0x0
    Load command 4
                cmd LC_DYLD_INFO_ONLY
            cmdsize 48
         rebase_off 1482752
        rebase_size 800
           bind_off 1483552
          bind_size 304
      weak_bind_off 0
     weak_bind_size 0
      lazy_bind_off 1483856
     lazy_bind_size 4600
         export_off 1488456
        export_size 7256
    Load command 5
         cmd LC_SYMTAB
     cmdsize 24
      symoff 1496840
       nsyms 1022
      stroff 1514820
     strsize 19152
    Load command 6
                cmd LC_DYSYMTAB
            cmdsize 80
          ilocalsym 0
          nlocalsym 470
         iextdefsym 470
         nextdefsym 346
          iundefsym 816
          nundefsym 206
             tocoff 0
               ntoc 0
          modtaboff 0
            nmodtab 0
       extrefsymoff 0
        nextrefsyms 0
     indirectsymoff 1513192
      nindirectsyms 407
          extreloff 0
            nextrel 0
          locreloff 0
            nlocrel 0
    Load command 7
              cmd LC_LOAD_DYLINKER
          cmdsize 32
             name /usr/lib/dyld (offset 12)
    Load command 8
          cmd LC_VERSION_MIN_MACOSX
      cmdsize 16
      version 10.12
          sdk 10.12
    Load command 9
          cmd LC_SOURCE_VERSION
      cmdsize 16
      version 0.0
    Load command 10
           cmd LC_MAIN
       cmdsize 24
      entryoff 102816
     stacksize 0
    Load command 11
              cmd LC_LOAD_DYLIB
          cmdsize 184
             name /nix/store/ayc7d1h0alm1y2pysr2hbv183g7d6ji7-x86_64-unknown-linux-gnu-binutils-2.31.1/x86_64-apple-darwin/x86_64-unknown-linux-gnu/lib/libbfd-2.31.1.dylib (offset 24)
       time stamp 2 Thu Jan  1 01:00:02 1970
          current version 0.0.0
    compatibility version 0.0.0
    Load command 12
              cmd LC_LOAD_DYLIB
          cmdsize 112
             name /nix/store/fxvxl64g1b336ayhzsrqdcv541zpb6lx-Libsystem-osx-10.12.6/lib/libSystem.B.dylib (offset 24)
       time stamp 2 Thu Jan  1 01:00:02 1970
          current version 1226.10.1
    compatibility version 1.0.0
    Load command 13
              cmd LC_LOAD_DYLIB
          cmdsize 96
             name /nix/store/ww5azqlhhp064brjqi7k2c6s30qsi0bg-zlib-1.2.11/lib/libz.dylib (offset 24)
       time stamp 2 Thu Jan  1 01:00:02 1970
          current version 1.2.11
    compatibility version 1.0.0
    Load command 14
              cmd LC_LOAD_DYLIB
          cmdsize 104
             name /nix/store/rb3m0xyx1l1jg2bl837w11jhm6ax1yrs-gettext-0.20.1/lib/libintl.8.dylib (offset 24)
       time stamp 2 Thu Jan  1 01:00:02 1970
          current version 10.6.0
    compatibility version 10.0.0
    Load command 15
              cmd LC_RPATH
          cmdsize 96
             path /nix/store/dbs6flb7c6rbvb3n4q4i6ad0nmhw1iaj-swift-corefoundation/Library/Frameworks (offset 12)
    Load command 16
          cmd LC_FUNCTION_STARTS
      cmdsize 16
      dataoff 1495712
     datasize 840
    Load command 17
          cmd LC_DATA_IN_CODE
      cmdsize 16
      dataoff 1496552
     datasize 288
</details>

**Comment by LnL7** [comment]
Great, looks like the flags are not relevant in that case. I'll make it conditional.

**Comment by LnL7** [comment]
Actually, I'm not sure how to conditionalize this.

**Comment by mroi** [comment]
I have not tried, this is just a guess: Would it work to set `NIX_x86_64_apple_darwin_LDFLAGS` instead of `NIX_LDFLAGS`?

**Comment by LnL7** [comment]
I don't think so, it shouldn't happen for both darwin->linux as well as linux->darwin and that would only handle one of those. It also seems brittle, the infiix salt is an implementation detail of the ld wrappers.

Perhaps somebody with more knowledge of the cross infrastructure has an idea on how to add this without polluting stdenv/generic with these details. This is kind of the problem I'm running into https://gist.github.com/LnL7/137a46fb81d03aeeeec96017f7a58fc1.

**Comment by mroi** [comment]
If I diagnose the problem correctly, only the Apple’s linker understands these flags, the GNU binutils do not. Do you think this is correct?
If so, wouldn’t it make more sense to add these linker flags to the `cctools-binutils-darwin-wrapper` package instead of `student`?

**Comment by euank** [comment]
Now that #83180 is merged, this should be fixed and can be closed, right?

**Comment by mroi** [comment]
I tried testing this, but the fix has been merged to `staging` and I am not sure what is needed for this to propagate to the `unstable` Nixpkgs channel. Trying to cross-compile on `staging` fails for me with errors unrelated to this issue here, so I have trouble confirming that the fix works. From all I have seen, I would assume it does, though.
